[ADD][product_properties_by_category]Settings up for product properties

hugho-ad opened the pull request:
[TASK#3652](https://www.vauxoo.com/#id=3652&view_type=form&model=project.task&menu_id=602&action=778)
[HU#888](https://www.vauxoo.com/#id=888&view_type=form&model=user.story) sprint 14

Add field selection to product.category to setting up the call for bids y product.tamplate

hugho-ad committed:
[ADD][product_properties_by_category]Settings up for product properties

hugho-ad committed:
Merge branch '8.0' of https://github.com/Vauxoo/addons-vauxoo into 8.0-product_properties_by_category-#vx3652-dev-hugho-ad

zaoral commented on product_properties_by_category/README.rst:
```diff
@@ -0,0 +1,7 @@
+New functionality to configurate the propierties of Product Category
+
+A new section called "Product Properties" in the Product Category
+where can configure “Call for Bids” of the product.
```
@hugho-ad please change the description content.
1. complete the line of the module title in line 2 to the same number of characters of the line 1. This need to match to be a valid rst syntax.
2. Change the description of the module with this:

```
This module add a new section called "Product Property" in the Product Category
form view to manage a group of common default values to be set at the Products
that belongs to the Category.

When you change the Internal Category of the Product the default values defined in
the Category for the Products will be set. This configuration is set but is not mandatory,
you can manually change the Products fields as you need.

For the current version, only works with the "Call For Bids" Product field.
When you update your Products Internal Category the "Call For Bids" field
in the Product will be updated with the default value given to the field
"Call For Bids" defined in the Product Category. If the Product Category
have not defined a default value for "Call For Bids" then will look into the
parent Product Category an so on until find a default value. If not default
value is set on any of the parent product category then the "Call For Bids"
field in Product will be set to False.
```

zaoral commented on product_properties_by_category/__openerp__.py:
```diff
@@ -0,0 +1,43 @@
+
+{
+    "name": "Product Properties by Category",
+    "summary": "Set Properties by Category",
```
@hugho-ad change to 

```
-   "summary": "Set Properties by Category",
+   "summary": "Set Product Default Properties from the Product Category",
```

zaoral commented on product_properties_by_category/i18n/es.po:
```diff
@@ -0,0 +1,52 @@
+#. module: product_properties_by_category
+#: selection:product.category,purchase_requisition:0
+msgid "Not Set"
+msgstr "No Configurado"
```
@hugho-ad "Sin Default". Apply the same to the es_PA translation

zaoral commented on product_properties_by_category/model/pruduct_category.py:
```diff
@@ -0,0 +1,62 @@
+        and update the state of the field "Call of Bids"
+        in the product.template (Boolean field with the same name).
+        If the field purchase_requisition product.category has "Not Set" option
+        then review the product category father
```
@hugho-ad change "product category father" for "product category parent"

nhomar replied:
In fact is: parent product catgory

zaoral commented on product_properties_by_category/model/pruduct_category.py:
```diff
@@ -0,0 +1,62 @@
+        self.purchase_requisition = self.check_parents_prch_rqs(self.categ_id)
+
+    @api.one
+    def check_parents_prch_rqs(self, categ_id):
```
@hugho-ad this method is correct. but is a recursion so will bring computational cost. There is a different way yo fullfil this issue in a linear cost. I will search for an example and I will reference to you to make the change.

zaoral commented:
@hugho-ad please change your file name model/pruduct_category.py. to model/product_category.py

hugho-ad committed:
[REF][product_properties_by_category]modify the description module, corrected translation

zaoral commented:
@hugho-ad please Hugo add a unit test to this module. Please use this schema:

```
1. Setup: Create o search a product that have a hierarchy product category.
    Product < Category A < Category B < Category C.
2. Test 01: Check that the product receive the default value defined in the product.category
2.1. Set "Call for Bids" False to the product.
2.2. Add a default value "Set True" for "Call for Bids" of product category first in line (Category A).
2.3. Simulate the onchange and check that the field in the product change for False to True.
3. Test 02: No restriction over the inherit values. The user can manually change a value inherit form the product.category as is needed.
3.1. Taking the same product in Test 01 change the "Call for Bids" in product to False.
3.2. Check the new value of product "call for bids" with the product.category A "call for bids" to check that they are different.
4. Test 03: Check pull the default value form the parent category one level above.
4.1. Remove the default value in Category A and add a default value "Set True" in Category B.
4.2. Simulate the onchane and check that de product "Call for Bids" change to default describe in Category B.
5. Test 04: What happen when is not default value defined in the product category or in either of its parents.
5.1. Change the product "Call for Bids" to True.
5.2. Remove all default in Category A/B/C. (Set then to "No set").
5.3. Simulate the onchange and check that the Call for Bids change form True to False.
```

zaoral commented:
@hugho-ad please add a string and help to the field ProductCategory.purchase_requistion, use the ones defined in the [product](https://github.com/vauxoo/odoo/blob/8.0-vauxoo/addons/purchase_requisition/purchase_requisition.py#L387). For this maybe you will need to update the translation files too.

zaoral committed:
[IMP] multiple improves:
- [ADD] modificate the name of the onchange of the Internal Category field
  to make it generic and mnemonic to add more product default properties in
  the future. Change the docstring of the method to indicate the generic
  behavior and results.
- [ADD] the method who search about the call for bids default value in the
  product category change from be a recursive method to a iterative method. A
  docstrings hrlp was added. Rename the method for mnemonic and delete the
  passed parameters. there was not needed.

zaoral commented:
@hugho-ad I create a pull request when you can find a diff of the changes I want you to test and apply to this PR vauxoo-dev/addons-vauxoo/pull/7, After finish the unit test please try to make merge into your branch.

zaoral committed:
Merge branch '8.0-product_properties_by_category-#vx3652-dev-hugho-ad' of github.com:vauxoo-dev/addons-vauxoo into 8.0-#vx3652-patch-kty

Conflicts:
	product_properties_by_category/model/product_category.py

zaoral committed:
[ADD] add index.html to add a web page representing the description of the new
module with youtube's link video with module's functionality.

hugho-ad committed:
[ADD][product_properties_by_category] added tests and modify the field purchase_requisition

hugho-ad committed:
Merge branch '7.0' of https://github.com/Vauxoo/addons-vauxoo into 8.0-product_properties_by_category-#vx3652-dev-hugho-ad

hugho-ad committed:
Merge pull request #7 from vauxoo-dev/8.0-#vx3652-patch-kty

[WIP] multiple improves

hugho-ad committed:
[IMP][product_properties_by_category]

hugho-ad committed:
Merge branch '8.0' of https://github.com/Vauxoo/addons-vauxoo into 8.0-product_properties_by_category-#vx3652-dev-hugho-ad

hugho-ad committed:
[FIX][product_properties_by_category] fix function get_purchase_requisition_default adding a break sentence, change tests functions

hugho-ad commented:
@zaoral ,I'm corrected, the function get_purchase_requisition_default by adding a break sentence
the tests was modified in order to your request.
Categoy A like the child category, category b like the parent and category C like parent of the parent

Regards

hugho-ad committed:
[IMP][product_properties_by_category] change decorator of get_purchase_requisition_default function, to retorn a bool and not a list

zaoral commented:
+1

zaoral commented:
@nhomar This PR is ready to review.

NOTE: The testing tools results: Runbot Green and we have a "False Red" for travis. This False Red is becuase a pylint error that currently is been in progress to be resolve for the vauxoo team.
